caliban_toolbox/reshape_data.py

```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
     log_path = os.path.join(save_dir, 'log_data.json')
     with open(log_path, 'w') as write_file:
         json.dump(log_data, write_file)
+
+
+def get_saved_file_path(dir_list, fov_name, row, col, slice, file_ext='.npz'):
+    """Helper function to identify correct file path for an npz file
+
+    Args:
+        dir_list: list of files in directory
+        fov_name: string of the current fov_name
+        row: int of current row
+        col: int of current col
+        slice: int of current slice
+        file_ext: extension file was saved with
+
+    Returns:
+        file_path: string of formatted file path with appropriate ending
+    """
+
+    base_string = 'fov_{}_row_{}_col_{}_slice_{}'.format(fov_name, row, col, slice)
+    string_matches = [string for string in dir_list if base_string + '_save_version' in string]
+
+    if len(string_matches) == 0:
+        full_string = base_string + file_ext
+    elif len(string_matches) == 1:
+        full_string = string_matches[0]
+    else:
+        raise ValueError('Multiple save versions found: '
+                         'please select only a single save version. {}'.format(string_matches))
+    return full_string
+
+
+def load_npzs(crop_dir, log_data, verbose=True):
+    """Reads all of the cropped images from a directory, and aggregates them into a single stack
+
+    Args:
+        crop_dir: path to directory with cropped npz or xarray files
+        log_data: dictionary of parameters generated during data saving
+
+        verbose: flag to control print statements
+
+    Returns:
+        stack: combined array of all labeled images
+    """
+
+    fov_names = log_data['fov_names']
+    num_crops, num_slices = log_data.get('num_crops', 1), log_data.get('num_slices', 1)
+    num_row_crops, num_col_crops = log_data.get('num_row_crops', 1), log_data.get('num_col_crops', 1)
+    fov_len, stack_len, _, _, row_size, col_size, _ = log_data['original_shape']
+    slice_stack_len = log_data.get('slice_stack_len', stack_len)
+
+    row_crop_size, col_crop_size = log_data.get('row_crop_size', row_size), log_data.get('col_crop_size', col_size)
+    save_format = log_data['save_format']
+
+    stack = np.zeros((fov_len, slice_stack_len, num_crops, num_slices, row_crop_size, col_crop_size, 1))
+    saved_files = os.listdir(crop_dir)
+
+    # loop through all npz files
+    for fov_idx, fov_name in enumerate(fov_names):
+        crop_idx = 0
+        for row in range(num_row_crops):
+            for col in range(num_col_crops):
+                for slice in range(num_slices):
```

Comment:
There must be a better way to do this. I am not wild about 4 nested for loops.

Reply:
The data structure that this is looping through is an array with (at an abstract level) the following format: [FOVs, x/y crops, z slices]. For each FOV, we have images from distinct x/y crops. For each x/y crop, we can have images from distinct z slices. 

We currently loop over the 'row' crops and 'column' crops, which is sort of a meaningless distinction. Those two loops could easily be collapsed into a singe loop over all crops. 

However, I'm not sure how to get below 3 nested loops. We really do have 3 different levels of data.

Reply:
I agree with Noah on this one. Big datasets get chopped up across each dimension to make them reasonable to annotate, and we end up with many smaller npzs to put back together. There's not a way for us to get around having fovs * rows * cols * slices number of npzs to load, and I think the very nested structure makes the most sense here, even if in general we try to avoid nested loops.

Reply:
I modified it to now have only 3 for loops. This actually cleaned up some of the other code as well.